### ISSUE TITLE
Local symbol inconsistencies

### ISSUE THREAD

**Issue by dbrotz** [rednex/rgbds#423]
Making local symbols with the `.` prefix was intended for labels, but the feature interacts with the other types of symbols in strange ways. In addition to labels, you can create constant symbols (with `EQU`/`SET`), string symbols (with `EQUS`), and macros.

Here's how the different types of symbols work currently.

### Labels
Neither creating nor referencing `.`-prefixed labels outside the scope of a normal label is allowed. Within the scope of a normal label, creating or referencing `.`-prefixed labels causes the name to be expanded by prefixing the name of the label in whose scope the `.`-prefixed label resides.

```
.test: ; error
jp .test ; error
foo:
.test: ; expands to foo.test
jp .test ; expands to foo.test
```

### Constant symbols
Creating a `.`-prefixed constant symbol never results in expansion. Referencing a `.`-prefixed constant symbol outside a scope doesn't result in expansion, but referencing one inside a scope does result in expansion.

```
.x = 1
printt "{d:.x}\n" ; prints 1
foo:
printt "{d:.x}\n" ; prints 0
foo.x = 2
printt "{d:.x}\n" ; prints 2
```

### String symbols
The rules are similar to those of constant symbols. However, there is a bug when referencing one inside a scope. The check to see if an identifier is a string symbol doesn't expand the label even inside a scope, but the label is expanded when retrieving the string symbol's value. This can result in a null pointer dereference.

```
.x EQUS "test"
foo:
.x ; null pointer dereference because foo.x doesn't exist
```

```
.x EQUS "test"
foo:
foo.x EQUS "printt \"OK\\n\""
.x ; prints OK
```
```
foo:
foo.x EQUS "printt \"OK\\n\""
.x ; prints nothing because .x is considered a label
```
### Macros
`.`-prefixed macro names are never expanded.

```
.x: MACRO
printt "test1\n"
ENDM
	.x ; prints "test1"

foo:
	.x ; prints "test1"

.y: MACRO
printt "test2\n"
ENDM

bar:
	.y ; prints "test2"
```

Clearly the most troubling case is the null pointer deference, but the whole system is inconsistent and confusing. I think that all symbols should follow the rules that labels do, except that only labels should be able to create scopes. That would break backward compatibility, but I find it hard to believe anyone was using `.` prefixes for symbols other than labels, especially due to how oddly it behaves.

**Comment by Ben10do** [comment]
I think these inconsistencies probably boil down to #181 focusing only on labels, without considering that constants or macros could also be defined locally. So, it’d be great to have this oversight addressed by, as you say, expanding the rules for labels to these other cases.

**Comment by ISSOtm** [comment]
I just think "."-prefixed names shouldn't be treated as special if they're not a symbol, as it doesn't make sense to scope them. At least not to have them scoped by labels.

**Comment by dbrotz** [comment]
What do you mean by names that aren't symbols?

**Comment by ISSOtm** [comment]
I meant "that aren't labels" >.<

**Comment by dbrotz** [comment]
The problem is that you can't tell if a symbol is a label or not before you look it up, so you would need to define some sort of resolution order, which is unnecessarily complex, in my opinion.

For example, in the following code, to handle the following input, you would first look up `foo.x` and check if it's a label. If so, you'd stop there. Otherwise, you'd look up `.x`.

```
foo:
db .x
```

I think that unifying the system so that you would just need to look up `foo.x` is conceptually simpler.

Why do you want to use a `.` prefix on other symbols but not have them be scoped?

**Comment by ISSOtm** [comment]
Another solution would be to ban non-label symbols starting with a dot, since they can be fairly confusing to the user and, well, *are* confusing to the parser.

**Comment by dbrotz** [comment]
That's something I was thinking of too. Like I said, I don't think anyone was using dot symbols for non-labels. Just disallowing the practice would solve the problem.

**Comment by ISSOtm** [comment]
Agreed on that.

**Comment by dbrotz** [comment]
There is one thing that would still be strange, though. Consider the following case. We have to expand `.x` to `foo.x` because it might be a label, but we would find the constant symbol `foo.x` instead. I suppose that if `foo.x` isn't a label, it could be considered an error.

```
foo.x = 1
foo:
db .x
```

**Comment by ISSOtm** [comment]
Well, that basically means that any scope expansion that takes place has to expand to a label, which makes sense, right?

**Comment by dbrotz** [comment]
Yeah. I think it can work.

**Comment by ISSOtm** [comment]
Fixed by #437.
